backend/src/crawler/crawler_base.py

```diff
@@ -0,0 +1,140 @@
+import abc
+from pydantic import AnyHttpUrl
+from tldextract import tldextract
+from sqlalchemy.orm import Session
+
+from .exceptions import DomainMismatchException
+
+from pydantic import BaseModel, Field, AnyHttpUrl
+
+
+class Headline(BaseModel):
+    title: str = Field(
+        default=...,
+        example="Title of the article",
+        description="The title of the article"
+    )
+    url: AnyHttpUrl | str = Field(
+        default=...,
+        example="https://www.example.com",
+        description="The URL of the article"
+    )
+
+
+class News(Headline):
+    time: str = Field(
+        default=...,
+        example="2021-10-01T00:00:00",
+        description="The time the article was published"
+    )
+    content: str = Field(
+        default=...,
+        example="Content of the article",
+        description="The content of the article"
+    )
+
+
+class NewsWithSummary(News):
+    summary: str = Field(
+        default=...,
+        example="Summary of the article",
+        description="The summary of the article"
+    )
+    reason: str = Field(
+        default=...,
+        example="Reason of the article",
+        description="The reason of the article"
+    )
+
+
+class NewsCrawlerBase(metaclass=abc.ABCMeta):
+    news_website_url: AnyHttpUrl | str
+    news_website_news_child_urls: list[AnyHttpUrl | str]
+
+    @abc.abstractmethod
+    def get_headline(
+            self, search_term: str, page: int | tuple[int, int]
+    ) -> list[Headline]:
+        """
+        Searches for news headlines on the news website based on a given search term and returns a list of headlines.
+
+        This method searches through the entire news_website_url using the specified search term, and returns a list
+        of Headline namedtuples, where each Headline includes the title and URL of a news article. The page parameter
+        can be an integer representing a single page number or a tuple representing a range of page numbers to search
+        through.
+        # The offset and limit parameters apply to the resulting list of headlines, allowing you to skip a
+        # certain number of headlines and limit the number of headlines returned, respectively.
+
+        :param search_term: A search term to search for news articles.
+        :param page: A page number (int) or a tuple of start and end page numbers (tuple[int, int]).
+        # :param offset: The number of headlines to skip from the beginning of the list.
+        # :param limit: The maximum number of headlines to return.
+        :return: A list of Headline namedtuple  s, each containing a title and a URL.
+        """
+        return NotImplemented
+
+    @abc.abstractmethod
+    def parse(self, url: AnyHttpUrl | str) -> News:
```

Comment:
Since this function takes a news url belonging to the news website, it should only be called by `validate_and_parse` method. As a result, we can rename it with an underscore, which means that the `_parse` is an internal method and should not be called externally.

```suggestion
    @abc.abstractmethod
    def _parse(self, url: AnyHttpUrl | str) -> News:
```